test/src/lib/backoff.spec.ts

```diff
@@ -0,0 +1,216 @@
+import { assert, expect } from 'chai';
+import * as sinon from 'sinon';
+
+import { StatusError } from '../../../src/lib/errors';
+import { withBackoff, OnFailureInfo } from '../../../src/lib/backoff';
+
+const DEFAULT_OPTIONS = {
+	maxRetries: 5,
+	maxDelay: 900000, // 15 minutes
+	minDelay: 10000, // 10 seconds
+};
+
+describe('lib/backoff', async () => {
+	let clock: sinon.SinonFakeTimers;
+
+	beforeEach(() => {
+		clock = sinon.useFakeTimers();
+	});
+
+	afterEach(() => {
+		clock.restore();
+	});
+
+	it('resolves after 3 retries', async () => {
+		// Create a function that will fail 3 times so it succeeds on the 4th
+		const failer = new Failer(3);
+		// Wrap function withBackoff
+		const fnWithBackoff = withBackoff(async () => {
+			return failer.willResolve('fails 3 times then resolves on 4th');
+		}, DEFAULT_OPTIONS);
+		// Call function and allow clock to trigger all events
+		clock.runAllAsync();
+		await expect(fnWithBackoff()).to.eventually.equal(
+			'fails 3 times then resolves on 4th',
+		);
+		// Check that function was called 4 times (failed 3 times, succeeds on 4th)
+		expect(failer.callCount).to.equal(4);
+	});
+
+	it('should not call the function before minDelay', async () => {
+		const failer = new Failer(3);
+		const minDelay = Math.floor(Math.random() * 1000);
+		const myBackoffFunc = withBackoff(
+			async () => {
+				return failer.willResolve('ok');
+			},
+			{
+				minDelay,
+			},
+		);
+		// Function should have been called 0 times to start
+		expect(failer.callCount).to.equal(0);
+		// Call function
+		myBackoffFunc();
+		// Check that function is ran atleast once
```

Comment:
```suggestion
		// Check that function was run at least once
```
nitpicky, feel free to ignore